### PR TITLE
Expire defaults when HTTP-date is not possible to parse (#20362) and  'If-Modified-Since' is more strict (excluding 0 - rfc 7232 3.3).

### PR DESCRIPTION
**Pull request by rbudzko**
> A cache recipient MUST interpret invalid date formats, especially the value "0", as representing a time in the past (i.e., "already expired").

and

> A recipient MUST ignore the If-Modified-Since header field if the received field-value is not a valid HTTP-date [...] .

### CODE REVIEW

**Comment by rbudzko** [comment]
Although I have some doubts about interpretation of If-Modified-Since. Can you share your opinions?

**Comment by ktoso** [comment]
Refs #20362

**Comment by ktoso** [comment]
OK TO TEST

**Comment by patriknw** [comment]
I have no clue

File: akka-http-core/src/main/scala/akka/http/impl/model/parser/CommonRules.scala [hunk 1]
```diff
@@ -61,12 +61,20 @@ private[parser] trait CommonRules { this: Parser with StringBuilding ⇒
   def `quoted-cpair` = `quoted-pair`
 
   // ******************************************************************************************
+  // http://tools.ietf.org/html/rfc7234#section-5.3
+  // ******************************************************************************************
+
+  def `expires-date`: Rule1[DateTime] = rule {
+    (`HTTP-date` | zeroOrMore(ANY) ~ push(DateTime.MinValue)) ~ OWS
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/CommonRules.scala]: Good

**Comment by ktoso** [comment]
Expires change surely is good.
If-Modified-Since I also tend to think it's good, but wording is a bit iffy indeed: https://tools.ietf.org/html/rfc7232#section-3.3

File: akka-http-core/src/main/scala/akka/http/impl/model/parser/CommonRules.scala [hunk 2]
```diff
@@ -61,12 +61,20 @@ private[parser] trait CommonRules { this: Parser with StringBuilding ⇒
...
+  }
+
+  // ******************************************************************************************
   // http://tools.ietf.org/html/rfc7231#section-7.1.1.1
   // but more lenient where we have already seen differing implementations in the field
   // ******************************************************************************************
 
   def `HTTP-date`: Rule1[DateTime] = rule {
-    (`IMF-fixdate` | `asctime-date` | '0' ~ push(DateTime.MinValue)) ~ OWS
+    (`IMF-fixdate` | `asctime-date`) ~ OWS
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/CommonRules.scala]: I think the change probably is good, but can we make it in a separate PR nevertheless?
So keep the `| '0'` in this commit, should also pass the tests of expires AFAICS, and then change this one separately?

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/CommonRules.scala]: Or nevermind actually... they do belong together in a way anyway

**Comment by ktoso** [comment]
I've read up on the RFCs and this seems valid, if @2beaucoup or @jrudolph have an opinion here it might be good to hear, if not I'll merge in a bit.

File: akka-http-core/src/main/scala/akka/http/impl/model/parser/CommonRules.scala (hunk 2 above)
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/CommonRules.scala]: The code-comment above suggests that this was intentional.

**Comment by rbudzko** [comment]
I've dig into history and looks like spray team decided to add 0 for a reason. For example:
> =htc allow `0` as special value `If-Modified-Since` and similar headers  Port of https://github.com/spray/spray/issues/942

Looks like (quick conversation with u.Google) if-modified-since, if-range, if-unmodified-since, last-modified like a company of '0' value (not by RFC but by real life). Expires is different though it seems to like '0', '-1' or even others - probably because later is described differently in RFC then earlier (explicit text about treating invalids as the past).

Maybe indeed as @ktoso proposed ` | '0' ~ push(DateTime.MinValue` should remain in power for HTTP-date so for all date headers apart of expires which will allow any value as fallback to the past, so `| zeroOrMore(ANY) ~ push(DateTime.MinValue)`.

**Comment by ktoso** [comment]
Yes please :)

**Comment by rbudzko** [comment]
Updated.

**Comment by ktoso** [comment]
Now LGTM :) Thanks a lot @rbudzko
